Python/pythonrun.c

```diff
@@ -1649,6 +1649,80 @@ PyOS_CheckStack(void)
 
 #endif /* WIN32 && _MSC_VER */
 
+#if defined(__APPLE__)
+
+#define MAC_OS_10_9 13
+#define MAC_OS_10_10 14
+#define MAC_OS_10_11 15
+
+#include <sys/utsname.h>
+
+size_t
+Get_Stack_Check_Size(const pthread_t *thread_self)
+{
+    size_t stack_space = 0;
+    if (pthread_main_np() == 1) {
+        // On macOS 10.09 - 10.11 pthread_get_stacksize_np
+        // returns wrong stack size on main thread.
+        // ref: https://github.com/rust-lang/rust/issues/43347#issuecomment-316783599
+        // ref: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8020753
+        struct utsname sysinfo;
+        uname(&sysinfo);
+        const int major_version = atoi(sysinfo.release);
+        if (major_version == MAC_OS_10_9  ||
+            major_version == MAC_OS_10_10 ||
+            major_version == MAC_OS_10_11) {
+            struct rlimit limit;
+            getrlimit(RLIMIT_STACK, &limit);
+            stack_space = limit.rlim_cur;
+        } else {
+            stack_space = pthread_get_stacksize_np(*thread_self);
+        }
+    } else {
+        stack_space = pthread_get_stacksize_np(*thread_self);
+    }
+    return stack_space;
+}
+
+void
+Init_Stack_Status(PyThreadState *tstate) {
+    if (tstate->stack_space == 0) {
+        const pthread_t thread_self = pthread_self();
+        size_t stack_space = Get_Stack_Check_Size(&thread_self);
+        tstate->stack_space = stack_space;
+        tstate->last_stack_remain = stack_space;
+    }
+}
+
+/*
+ * Return non-zero when we run out of memory on the stack; zero otherwise.
+ */
+int
+PyOS_CheckStack(void)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    Init_Stack_Status(tstate);
```

Comment:
Micro-optimization: I would suggest to check `if (tstate->stack_space == 0) {` here.